### PR TITLE
Generalize few watch tests from cacher

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #109831

```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go [hunk 1]
```diff
@@ -94,6 +100,24 @@ func testWatch(ctx context.Context, t *testing.T, store storage.Interface, recur
 				return nil, fields.Set{"spec.nodeName": pod.Spec.NodeName}, nil
 			},
 		},
+	}, {
+		name:       "filtering",
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:104]: FTR - changes in this test are effectively covering what was previously tested with `TestFiltering` test.

File: staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go [hunk 2]
```diff
@@ -226,6 +272,60 @@ func RunTestWatchFromNoneZero(ctx context.Context, t *testing.T, store storage.I
...
+func RunTestDelayedWatchDelivery(ctx context.Context, t *testing.T, store storage.Interface) {
+	_, storedObj := testPropagateStore(ctx, t, store, &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "test-ns"}})
+	startRV := storedObj.ResourceVersion
+
+	watcher, err := store.Watch(ctx, "/pods/test-ns", storage.ListOptions{ResourceVersion: startRV, Predicate: storage.Everything, Recursive: true})
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+
+	// Depending on the implementation, different number of events that
```
Comment (stevekuznetsov) [inline staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:284]: Would it be useful to make these implementation-specific counts part of the `RunTest...` signature?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go:284]: That was the case for my original implementation. But for etcd this number doesn't really exist at all, as watches are independent on etcd.
So to make the tests as simple as possible (given that we don't have any reasonable number to put here for etcd), I would leave as is, until at least we have another implementation of storage interface.

**Review by wojtek-t — COMMENTED** [review]
@stevekuznetsov - thanks for your comment, PTAL

**Comment by stevekuznetsov** [comment]
Thanks for the response, makes sense.

/lgtm
/approve

**Pull request merged**